builtin/rebase--helper.c

```diff
@@ -10,22 +10,36 @@ static GIT_PATH_FUNC(path_squash_onto, "rebase-merge/squash-onto")
 
 static int get_revision_ranges(const char *upstream, const char *onto,
 			       const char **head_hash,
-			       char **revisions)
+			       char **revisions, char **shortrevisions)
 {
+	const char *shortupstream, *shorthead;
 	struct object_id orig_head;
 
 	if (get_oid("HEAD", &orig_head))
 		return error(_("no HEAD?"));
 
 	*head_hash = find_unique_abbrev(&orig_head, GIT_MAX_HEXSZ);
 
+	if (shortrevisions)
+		shorthead = find_unique_abbrev(&orig_head, DEFAULT_ABBREV);
+
 	if (upstream) {
 		struct object_id upstream_oid;
 
 		get_oid(upstream, &upstream_oid);
-		*revisions = xstrfmt("%s...%s", upstream, *head_hash);
-	} else
-		*revisions = xstrfmt("%s...%s", onto, *head_hash);
+
+		if (revisions)
+			*revisions = xstrfmt("%s...%s", upstream, *head_hash);
+		if (shortrevisions) {
+			shortupstream = find_unique_abbrev(&upstream_oid, DEFAULT_ABBREV);
+			*shortrevisions = xstrfmt("%s..%s", shortupstream, shorthead);
+		}
+	} else {
+		if (revisions)
+			*revisions = xstrfmt("%s...%s", onto, *head_hash);
+		if (shortrevisions)
+			*shortrevisions = xstrndup(shorthead, strlen(shorthead));
```

Comment:
Why not just `*shortrevisions = xstrdup(shorthead)`?

Reply:
Wait, is that even pointing out the right issue? Since `revisions` is essentially `$onto..$head_hash`, `shortrevisions` should be `$short_onto..$short_head`.

/me goes and looks

Hmpf. The code agrees: https://github.com/git/git/blob/v2.18.0/git-rebase--interactive.sh#L898

I think this is a bug in `git-rebase--interactive.sh`...

Reply:
As far as I can see, `$shortrevisions` was introduced with d911d1465d ([see here](https://github.com/git/git/commit/d911d1465d134cbd126e531a544f33b13f4e855a#diff-0f15aff45d5dd346465c35597a5f274eR659)), and it was already like that here. Changing its value to `$short_onto..$short_head` does not seem to break any test, though.

Reply:
It should be a strong hint that *I myself*, during a review, stumbled over this. So please do expect the same question to come up on the Git mailing list. And the best way to prepare for this is to either dig into history to find out whether there is any good reason for this, or whether this is a bug. In the latter case, fix it. In the former, document it in the commit message. That's what commit messages are for, after all, to answer important question about the patch such as this one, that are not readily answered by merely looking at the diff.

Reply:
[Well, you tried to change this a while ago](https://public-inbox.org/git/alpine.DEB.1.00.0901260031460.14855@racer/). Apart from that, I did not really found something interesting on the list.

Also, I wanted to point out that this specific code is executed when the user call a `rebase --root` (empty upstream means `rebase --root`), [and that `$shortonto` is only used in the help message](https://github.com/agrn/git/blob/ag/init-revisions-c/sequencer.c#L4512-L4515). So, if I make a `rebase -i --root`, I have something like that:

```
pick f203597 First commit
pick 7855fc1 changes
…
pick 5488a93 more changes
pick fb6bb71 last commit

# Rebase fb6bb71 onto c9e8c18 (447 commands)
#        ^            ^~~~~~~ $shortonto
#        ~~~~~~~ $shortrevisions
```

I find this a bit unclear.

Reply:
"a while ago"... *snickers*... 9.5 years ago. I did not remember, I must confess.

But now I understand why `shortrevisions` is defined that way: in the `--root` case, there is no `..` way to specify the commit range, sadly. The `$onto..$head`, as `revisions` does it, is actually incorrect (although it does not matter in practice, as nobody would use an `onto` that is reachable from `head` in the `--root` case: that would cause essentially a non-`--root` rebase).

Still, I feel uneasy about all of this discussion *not* being distilled into a small, clarifying paragraph in the commit message. The issue *is* confusing, after all.